File: java/commons/src/test/java/co/com/bancolombia/utils/DelayedTaskRunnerTest.java
package co.com.bancolombia.utils;

import org.assertj.core.data.Offset;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class DelayedTaskRunnerTest {

    @Test
    public void shouldProcessForADelayTime() {
        // Arrange
        long delay = 1000;
        long expected = System.currentTimeMillis() + delay;
        // Act
        int times = DelayedTaskRunner.start(() -> 10f / 50f, Percentage.MAX_PERCENTAGE, delay);
        // Assert
        long realEnd = System.currentTimeMillis();
        assertThat(realEnd).isCloseTo(expected, Offset.offset(50L));
        assertThat(times).isGreaterThan(0);
        System.out.println("Real delay: " + (realEnd - (expected - delay)) + " expectedDelay: " + delay);
    }
}

File: nodejs/commons/lib/delayed-task-runner.test.js
const { start, startReactive } = require('./delayed-task-runner');


test('should delay a task', () => {
    // Arrange
    const delay = 1000;
    let expected = (new Date()).getTime() + delay;
    // Act
    start(() => 10 / 50, 100, delay)
        .then(times => {
            // Assert
            const realEnd = (new Date()).getTime();
            // console.log("Real delay: " + (realEnd - (expected - delay)) + " expectedDelay: " + delay);
            expect(realEnd).toBeCloseTo(expected, -100)
            expect(times).toBeGreaterThan(0);
        });
});


test('should delay a task with Observable', done => {
    // Arrange
    const delay = 1000;
    let expected = (new Date()).getTime() + delay;
    // Act
    const fn = async () => Promise.resolve(10 / 50);

    startReactive(fn, 100, delay)
        .subscribe(times => {
            const realEnd = (new Date()).getTime();
            expect(realEnd).toBeCloseTo(expected, -100)
            expect(times).toBeGreaterThan(0);
            done();
        });

}, 5000);
